torch/csrc/distributed/rpc/ProcessGroupAgent.cpp

```diff
@@ -106,11 +102,9 @@ void ProcessGroupAgent::sync() {
   // the lock below, because other processes might not enter sync() until it
   // gets some response from this RpcAgent.
   pg_->barrier()->wait();
-  // Acquire the lock on the send queue to prevent additional messages to be put
-  // onto the send queue.
-  std::unique_lock<std::mutex> lock(sendQueueMutex_);
-  // Wait until the send queue is depleted.
-  workConsumeCV_.wait(lock, [&] { return sendQueue_.empty(); });
+  // Wait until the all send works are done.
+  // NB: There might be additional send works inserted while waiting.
```

Comment:
Why is that? Wouldn't `::sync` only be called by the same user thread that ends up issuing the sends?

Reply:
Yes, but there could be nested RPC calls. Say when calling sync on A, A has one pending send to B, and when running that RPC on B, it triggers B->A->B. In this case one additional send will be inserted into A's task queue.